File: 20.valid-parentheses/solution.ts
function isValid(s: string): boolean {
    const stack: string[] = [];
    const pairs = {
        "}": "{",
        ")": "(",
        "]": "[",
    };
    for (const char of s) {
        if (!(char in pairs)) {
            stack.push(char);
            continue;
        }
        const top = stack.pop()!;
        if (top !== pairs[char]) {
            return false;
        }
    }
    return stack.length === 0;
};

File: 20.valid-parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {
            "[": "]",
            "(": ")",
            "{": "}"
        }

        stack = []
        for char in s:
            if char in pairs:
                stack.append(pairs[char])
                continue
            if stack and stack[-1] == char:
                stack.pop()
                continue
            return False
        return len(stack) == 0
